build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/transport/GenerateTransportVersionDefinitionTask.java

```diff
@@ -0,0 +1,236 @@
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.gradle.internal.transport;
+
+import org.gradle.api.DefaultTask;
+import org.gradle.api.file.ConfigurableFileCollection;
+import org.gradle.api.provider.Property;
+import org.gradle.api.services.ServiceReference;
+import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.InputFiles;
+import org.gradle.api.tasks.Optional;
+import org.gradle.api.tasks.PathSensitive;
+import org.gradle.api.tasks.PathSensitivity;
+import org.gradle.api.tasks.TaskAction;
+import org.gradle.api.tasks.options.Option;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * This task generates transport version definition files. These files
+ * are runtime resources that TransportVersion loads statically.
+ * They contain a comma separated list of integer ids. Each file is named the same
+ * as the transport version name itself (with the .csv suffix).
+ */
+public abstract class GenerateTransportVersionDefinitionTask extends DefaultTask {
+
+    /**
+     * Files that contain the references to the transport version names.
+     */
+    @InputFiles
+    @PathSensitive(PathSensitivity.RELATIVE)
+    public abstract ConfigurableFileCollection getReferencesFiles();
+
+    @ServiceReference("transportVersionResources")
+    abstract Property<TransportVersionResourcesService> getResourceService();
+
+    @Input
+    @Optional
+    @Option(option = "name", description = "The name of the Transport Version definition, e.g. --name=my_new_tv")
+    public abstract Property<String> getTransportVersionName();
+
+    @Input
+    @Optional
+    @Option(
+        option = "backport-branches",
+        description = "The branches this definition will be backported to, e.g. --backport-branches=9.1,8.19"
+    )
+    public abstract Property<String> getBackportBranches();
+
+    @Input
+    @Optional
+    @Option(option = "increment", description = "The amount to increment the primary id for the main releaseBranch")
+    public abstract Property<Integer> getPrimaryIncrement();
+
+    /**
+     * The version number currently associated with the `main` branch, (e.g. as returned by
+     * VersionProperties.getElasticsearchVersion()).
+     */
+    @Input
+    public abstract Property<String> getMainReleaseBranch();
+
+    @TaskAction
+    public void run() throws IOException {
+        TransportVersionResourcesService resources = getResourceService().get();
+        Set<String> referencedNames = TransportVersionReference.collectNames(getReferencesFiles());
+        List<String> changedDefinitionNames = resources.getChangedReferableDefinitionNames();
+        String name = getTargetName(resources, referencedNames, changedDefinitionNames);
+
+        List<TransportVersionUpperBound> mainUpperBounds = resources.getUpperBoundsFromMain();
```

Comment:
The term `main` is overloaded and I think when many people read this (as I did) they initially think the branch, not `src/main`. Do we need this qualifier here? Generally we qualify source or resources with "main" when we want to distinguish them from "test" but that's not the case here. I'd be included to remove the `main` qualifier from everything in `TransportVersionResourcesService`. I think it's just more confusing than anything else.

Reply:
It is actually the branch! These methods read a given resource file _from the main branch_. The reason is that we want to get the "pristine" state of the file from main, so we can compare it to what the current branch has. This way it doesn't matter what garbage exists locally, we can always (idempotently) generate a new version, and get the resource files back into a valid state.

Reply:
Ah I see. Yeah, my only concern here is that, although we don't really intend for folks to run this locally, they might, and _their_ version of `main` might be out of date. I guess worst case there is they push their branch, open a PR, and we either catch that issue or remediate it for them?